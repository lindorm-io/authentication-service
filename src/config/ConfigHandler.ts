import { ConfigurationBase, IConfigurationDataBase, IConfigurationOptions } from "@lindorm-io/koa-config";

export interface IConfigurationData extends IConfigurationDataBase {
  SERVER_PORT: number;
  HOST: string;

  // Secrets
  CRYPTO_AES_SECRET: string;
  CRYPTO_SHA_SECRET: string;

  // Tokens
  JWT_ISSUER: string;

  // Authentication
  ACCOUNT_OTP_ISSUER: string;

  // Emails
  MAILGUN_API_KEY: string;
  MAILGUN_DOMAIN: string;
  MAILGUN_FROM: string;

  // Expiry
  JWT_ACCESS_TOKEN_EXPIRY: string;
  JWT_AUTHORIZATION_TOKEN_EXPIRY: string;
  JWT_IDENTITY_TOKEN_EXPIRY: string;
  JWT_MULTI_FACTOR_TOKEN_EXPIRY: string;
  JWT_REFRESH_TOKEN_EXPIRY: string;

  // Infrastructure
  REDIS_PORT: number;

  MONGO_INITDB_ROOT_USERNAME: string;
  MONGO_INITDB_ROOT_PASSWORD: string;
  MONGO_HOST: string;
  MONGO_EXPOSE_PORT: number;
  MONGO_DB_NAME: string;

  // Services
  DEVICE_SERVICE_BASE_URL: string;
  DEVICE_SERVICE_AUTH_USERNAME: string;
  DEVICE_SERVICE_AUTH_PASSWORD: string;

  IDENTITY_SERVICE_BASE_URL: string;
  IDENTITY_SERVICE_AUTH_USERNAME: string;
  IDENTITY_SERVICE_AUTH_PASSWORD: string;
}

export class ConfigHandler extends ConfigurationBase<IConfigurationData> {
  constructor(options: IConfigurationOptions<IConfigurationData>) {
    super(options);
  }
}
