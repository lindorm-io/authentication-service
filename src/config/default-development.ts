import { IConfiguration } from "../typing";
import { NodeEnvironment } from "@lindorm-io/core";

export const configuration: IConfiguration = {
  NODE_ENVIRONMENT: NodeEnvironment.DEVELOPMENT,
  SERVER_PORT: 3000,
  HOST: "http://localhost/",

  JWT_ISSUER: "https://dev.lindorm.io/",
  JWT_ACCESS_TOKEN_EXPIRY: "1 minutes",
  JWT_AUTHORIZATION_TOKEN_EXPIRY: "30 minutes",
  JWT_IDENTITY_TOKEN_EXPIRY: "5 minutes",
  JWT_MULTI_FACTOR_TOKEN_EXPIRY: "30 minutes",
  JWT_REFRESH_TOKEN_EXPIRY: "5 minutes",

  CRYPTO_AES_SECRET: "secret",
  CRYPTO_SHA_SECRET: "secret",

  ACCOUNT_OTP_ISSUER: "dev.lindorm.io",

  MAILGUN_API_KEY: "api-key",
  MAILGUN_DOMAIN: "dev.lindorm.io",
  MAILGUN_FROM: "noreply@lindorm.io",

  MONGO_INITDB_ROOT_USERNAME: "root",
  MONGO_INITDB_ROOT_PASSWORD: "password",
  MONGO_HOST: "localhost",
  MONGO_EXPOSE_PORT: 27017,
  MONGO_DB_NAME: "authentication",
};
