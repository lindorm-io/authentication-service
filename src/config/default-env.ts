import { config } from "dotenv";
import { isString } from "lodash";
import { IConfiguration } from "../typing";
import { NodeEnvironment } from "@lindorm-io/core";

config();

export const configuration: IConfiguration = {
  NODE_ENVIRONMENT: process.env.NODE_ENV || NodeEnvironment.DEVELOPMENT,
  SERVER_PORT: isString(process.env.SERVER_PORT) ? parseInt(process.env.SERVER_PORT, 10) : null,
  HOST: process.env.HOST,

  JWT_ISSUER: process.env.JWT_ISSUER,
  JWT_ACCESS_TOKEN_EXPIRY: process.env.JWT_ACCESS_TOKEN_EXPIRY,
  JWT_AUTHORIZATION_TOKEN_EXPIRY: process.env.JWT_AUTHORIZATION_TOKEN_EXPIRY,
  JWT_IDENTITY_TOKEN_EXPIRY: process.env.JWT_IDENTITY_TOKEN_EXPIRY,
  JWT_MULTI_FACTOR_TOKEN_EXPIRY: process.env.JWT_MULTI_FACTOR_TOKEN_EXPIRY,
  JWT_REFRESH_TOKEN_EXPIRY: process.env.JWT_REFRESH_TOKEN_EXPIRY,

  CRYPTO_AES_SECRET: process.env.CRYPTO_AES_SECRET,
  CRYPTO_SHA_SECRET: process.env.CRYPTO_SHA_SECRET,

  ACCOUNT_OTP_ISSUER: process.env.ACCOUNT_OTP_ISSUER,

  MAILGUN_API_KEY: process.env.MAILGUN_API_KEY,
  MAILGUN_DOMAIN: process.env.MAILGUN_DOMAIN,
  MAILGUN_FROM: process.env.MAILGUN_FROM,

  MONGO_INITDB_ROOT_USERNAME: process.env.MONGO_INITDB_ROOT_USERNAME,
  MONGO_INITDB_ROOT_PASSWORD: process.env.MONGO_INITDB_ROOT_PASSWORD,
  MONGO_HOST: process.env.MONGO_HOST,
  MONGO_EXPOSE_PORT: isString(process.env.MONGO_EXPOSE_PORT) ? parseInt(process.env.MONGO_EXPOSE_PORT, 10) : null,
  MONGO_DB_NAME: process.env.MONGO_DB_NAME,
};
