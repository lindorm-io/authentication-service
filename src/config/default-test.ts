import { IConfiguration } from "../typing";
import { NodeEnvironment } from "@lindorm-io/core";

export const configuration: IConfiguration = {
  NODE_ENVIRONMENT: NodeEnvironment.DEVELOPMENT,
  SERVER_PORT: 3000,
  HOST: "http://localhost/",

  JWT_ISSUER: "https://test.lindorm.io/",
  JWT_ACCESS_TOKEN_EXPIRY: "2 minutes",
  JWT_AUTHORIZATION_TOKEN_EXPIRY: "45 minutes",
  JWT_IDENTITY_TOKEN_EXPIRY: "1 days",
  JWT_MULTI_FACTOR_TOKEN_EXPIRY: "30 minutes",
  JWT_REFRESH_TOKEN_EXPIRY: "1 days",

  CRYPTO_AES_SECRET: "secret",
  CRYPTO_SHA_SECRET: "secret",

  ACCOUNT_OTP_ISSUER: "test.lindorm.io",

  MAILGUN_API_KEY: "api-key",
  MAILGUN_DOMAIN: "test.lindorm.io",
  MAILGUN_FROM: "noreply@lindorm.io",

  MONGO_INITDB_ROOT_USERNAME: "root",
  MONGO_INITDB_ROOT_PASSWORD: "password",
  MONGO_HOST: "localhost",
  MONGO_EXPOSE_PORT: 27017,
  MONGO_DB_NAME: "authentication",
};
