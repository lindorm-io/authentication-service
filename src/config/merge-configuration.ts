import { IConfiguration } from "../typing";
import { TObject } from "@lindorm-io/core";

export const mergeConfiguration = (primary: IConfiguration, secondary: IConfiguration): IConfiguration => {
  const object: TObject<any> = secondary;

  for (const [key, value] of Object.entries(primary)) {
    if (value === null || value === undefined) continue;
    object[key] = value;
  }

  const result: unknown = object;
  return result as IConfiguration;
};
