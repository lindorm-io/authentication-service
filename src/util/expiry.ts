import { add } from "date-fns";
import { stringToDurationObject } from "@lindorm-io/core";

export const getExpiryDate = (expiry: string): Date => {
  return add(Date.now(), stringToDurationObject(expiry));
};
