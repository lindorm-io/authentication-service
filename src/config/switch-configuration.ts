import { IConfiguration } from "../typing";
import { NodeEnvironment } from "@lindorm-io/core";
import { configuration as development } from "./default-development";
import { configuration as production } from "./default-production";
import { configuration as staging } from "./default-staging";
import { configuration as test } from "./default-test";

export const switchConfiguration = (environment: string): IConfiguration => {
  switch (environment) {
    case NodeEnvironment.PRODUCTION:
      return production;

    case NodeEnvironment.STAGING:
      return staging;

    case NodeEnvironment.DEVELOPMENT:
      return development;

    case NodeEnvironment.TEST:
      return test;

    default:
      return development;
  }
};
