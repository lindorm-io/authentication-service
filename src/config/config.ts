import { Audience } from "../enum";
import { IConfiguration } from "../typing";
import { configuration as env } from "./default-test";
import { mergeConfiguration } from "./merge-configuration";
import { switchConfiguration } from "./switch-configuration";
import { MongoConnectionType } from "@lindorm-io/mongo";
import { RedisConnectionType } from "@lindorm-io/redis";

export const { NODE_ENVIRONMENT } = env;

const config: IConfiguration = mergeConfiguration(env, switchConfiguration(NODE_ENVIRONMENT));

export const SERVER_PORT = config.SERVER_PORT;
export const HOST = config.HOST;

export const JWT_ISSUER = config.JWT_ISSUER;
export const JWT_ACCESS_TOKEN_EXPIRY = config.JWT_ACCESS_TOKEN_EXPIRY;
export const JWT_AUTHORIZATION_TOKEN_EXPIRY = config.JWT_AUTHORIZATION_TOKEN_EXPIRY;
export const JWT_IDENTITY_TOKEN_EXPIRY = config.JWT_IDENTITY_TOKEN_EXPIRY;
export const JWT_MULTI_FACTOR_TOKEN_EXPIRY = config.JWT_MULTI_FACTOR_TOKEN_EXPIRY;
export const JWT_REFRESH_TOKEN_EXPIRY = config.JWT_REFRESH_TOKEN_EXPIRY;

export const MAIL_HANDLER_CONFIG = {
  apiKey: config.MAILGUN_API_KEY,
  domain: config.MAILGUN_DOMAIN,
  environment: NODE_ENVIRONMENT,
  from: config.MAILGUN_FROM,
};

export const CRYPTO_PASSWORD_OPTIONS = {
  aesSecret: config.CRYPTO_AES_SECRET,
  shaSecret: config.CRYPTO_SHA_SECRET,
};

export const CRYPTO_SECRET_OPTIONS = {
  aesSecret: config.CRYPTO_AES_SECRET,
  shaSecret: config.CRYPTO_SHA_SECRET,
};

export const OTP_HANDLER_OPTIONS = {
  issuer: config.ACCOUNT_OTP_ISSUER,
  secret: config.CRYPTO_AES_SECRET,
};

export const BEARER_TOKEN_MW_OPTIONS = {
  issuer: config.JWT_ISSUER,
  audience: Audience.ACCESS,
};

export const MONGO_MW_OPTIONS = {
  type: MongoConnectionType.STORAGE,
  auth: {
    user: config.MONGO_INITDB_ROOT_USERNAME,
    password: config.MONGO_INITDB_ROOT_PASSWORD,
  },
  url: {
    host: config.MONGO_HOST,
    port: config.MONGO_EXPOSE_PORT,
  },
  databaseName: config.MONGO_DB_NAME,
};

export const REDIS_MW_OPTIONS = {
  type: RedisConnectionType.CACHE,
  port: config.REDIS_PORT,
};

export const TOKEN_ISSUER_MW_OPTIONS = {
  issuer: config.JWT_ISSUER,
};
