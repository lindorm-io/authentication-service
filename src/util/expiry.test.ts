import MockDate from "mockdate";
import { config } from "../config";
import { getExpiryDate } from "./expiry";

jest.mock("uuid", () => ({
  v4: jest.fn(() => "be3a62d1-24a0-401c-96dd-3aff95356811"),
}));

MockDate.set("2020-01-01 08:00:00.000");

describe("getExpiryDate", () => {
  test("should return a modified date for authorization token", () => {
    expect(getExpiryDate(config.JWT_AUTHORIZATION_TOKEN_EXPIRY)).toMatchSnapshot();
  });

  test("should return a modified date for refresh token", () => {
    expect(getExpiryDate(config.JWT_REFRESH_TOKEN_EXPIRY)).toMatchSnapshot();
  });
});
