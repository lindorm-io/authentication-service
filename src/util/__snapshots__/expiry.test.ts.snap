// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getExpiryDate should return a modified date for authorization token 1`] = `2020-01-01T07:45:00.000Z`;

exports[`getExpiryDate should return a modified date for refresh token 1`] = `2020-01-02T07:00:00.000Z`;
